composer/algorithms/swa/swa.py

```diff
@@ -108,36 +116,26 @@ def apply(self, event: Event, state: State, logger: Logger) -> None:
         """
         assert state.model is not None, 'We cannot apply SWA to None'
 
-        swa_start_epochs = int(self.hparams.swa_start * state.max_epochs)
+        swa_start_steps = state.step >= int(self.hparams.swa_start * state.max_epochs * state.steps_per_epoch)
 
         if event == Event.TRAINING_START:
-            self.swa_model = AveragedModel(state.model)
-
-        if event == Event.EPOCH_END and state.epoch == swa_start_epochs:
-            assert self.swa_scheduler is None, "SWA Scheduler should only be set once. Another algorithm "
-            "may have adjusted the max_epochs."
+            self.swa_model = AveragedModel(state.model.module)
+            log.info("Creating AveragedModel")
 
-            if self.hparams.swa_lr is None:
-                last_lr = state.schedulers.schedulers[0].get_last_lr()  # assumes ComposedScheduler
-                log.info(f'Setting SWA LR to {last_lr}')
-                self.hparams.swa_lr = last_lr
+        if self.hparams.swa_lr is None:
+            log.info(f'Setting SWA LR to default lr')
 
-            self.swa_scheduler = SWALR(
-                state.optimizers[0] if isinstance(state.optimizers, tuple) else state.optimizers,
-                swa_lr=self.hparams.swa_lr,
-                anneal_epochs=self.hparams.anneal_epochs,
-                anneal_strategy='cos',
-            )
-
-        if event == Event.EPOCH_END and state.epoch >= swa_start_epochs:
+        if swa_start_steps:
             self.swa_model.update_parameters(state.model)
-
-            if self.swa_scheduler is None:
-                raise ValueError('SWA LR scheduler was not set.')
-            self.swa_scheduler.step()
-
-        ## end of training
-        if event == Event.TRAINING_END:
-            update_bn(state.train_dataloader, self.swa_model)
-            state.model = self.swa_model
-            log.info('Updated BN and set model to the averaged model')
+            log.info("updating AveragedModel")
+
+            if state.max_epochs == 1:
+                update_bn(state.train_dataloader, self.swa_model)
+                state.model.module = self.swa_model.module
```

Comment:
I'm a bit concerned about how this line would work with DDP...curious if we tested it? Following line 129, would `state.model.module.update_parameters(self.swa_model)` work?